# Wrangle Pisa 2012
#### By Gabriela Sikora


### Abstract



### Table of Contents

<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#wrangling">Data Wrangling</a></li>
    <ul>
    <li><a href="#gathering">Gathering Data</a></li>
    <li><a href="#assessing1">Initial Assessment</a></li>
        <ul>
          <li><a href="#issues1">Identified Issues</a></li>
        </ul>
    <li><a href="#cleaning1">Cleaning Data</a></li>
    <li><a href="#assessing2">Final Assessment</a></li>
    </ul>
<li><a href="#sav">Store</a></li>
<li><a href="#conclusion">Conclusion</a></li>
<li><a href="#references">References</a></li>    
</ul>

<a id='intro'></a>
## Introduction



To begin, we first need to import all the relevant libraries:
#### Import

In [1]:
# import all packages
import numpy as np
import pandas as pd

<a id='wrangling'></a>
## Data Wrangling

In this section of the report, we will load in the data, check it for cleanliness, and then trim and clean the datasets for analysis. 

<a id='gathering'></a>
### Gathering Data

Unfortunately the Pisa dataset is too large to be uploaded to github, but if you would like to work with it yourself then place follow this link: https://s3.amazonaws.com/udacity-hosted-downloads/ud507/pisa2012.csv.zip 

In [2]:
# Encoding help found here: https://stackoverflow.com/questions/30462807/encoding-error-in-panda-read-csv

# Load in the Pisa dataframe
pisa = pd.read_csv('pisa2012.csv', encoding="cp1252", sep=',')

# Load in the Pisa dictionary
pisa_dict = pd.read_csv('pisadict2012.csv', encoding="cp1252", sep=',')

/Users/gabriela/Desktop/Code/myenv/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3044: DtypeWarning: Columns (15,16,17,21,22,23,24,25,26,30,31,36,37,45,65,123,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,284,285,286,287,288,289,290,291,292,293,294,295,296,297,298,299,300,301,302,303,307,308,309,310,311,312,313,314,315,316,317,318,319,320,321,322,323,324,325,326,327,328,329,330,331,332,333,334,335,336,337,338,339,340,341,342,343,344,345,346,347,348,349,350,351,352,353,354,355,356,357,376,377,378,379,380,381,382,383,384,385,386,387,388,389,390,391,392,393,394,395,396,397,398,399,400,401,402,403,475) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


<a id='assessing1'></a>
### First Assessment

In this section, we will do the preliminary visual and programmatic assessment of the pisa datasets to determine whether or not it holds any major quality or tidiness issues.

In [3]:
pisa

Unnamed: 0      CNT  SUBNATIO  STRATUM      OECD         NC  SCHOOLID  \
0                1  Albania     80000  ALB0006  Non-OECD    Albania         1   
1                2  Albania     80000  ALB0006  Non-OECD    Albania         1   
2                3  Albania     80000  ALB0006  Non-OECD    Albania         1   
3                4  Albania     80000  ALB0006  Non-OECD    Albania         1   
4                5  Albania     80000  ALB0006  Non-OECD    Albania         1   
5                6  Albania     80000  ALB0006  Non-OECD    Albania         1   
6                7  Albania     80000  ALB0006  Non-OECD    Albania         1   
7                8  Albania     80000  ALB0006  Non-OECD    Albania         1   
8                9  Albania     80000  ALB0006  Non-OECD    Albania         1   
9               10  Albania     80000  ALB0005  Non-OECD    Albania         2   
10              11  Albania     80000  ALB0005  Non-OECD    Albania         2   
11              12  Albania     80000  ALB0005  Non-OECD    Albania         2   
12              13  Albania     80000  ALB0005  Non-OECD    Albania         2   
13              14  Albania     80000  ALB0005  Non-OECD    Albania         2   
14              15  Albania     80000  ALB0005  Non-OECD    Albania         2   
15              16  Albania     80000  ALB0005  Non-OECD    Albania         2   
16              17  Albania     80000  ALB0005  Non-OECD    Albania         2   
17              18  Albania     80000  ALB0005  Non-OECD    Albania         2   
18              19  Albania     80000  ALB0005  Non-OECD    Albania         2   
19              20  Albania     80000  ALB0005  Non-OECD    Albania         2   
20              21  Albania     80000  ALB0005  Non-OECD    Albania         2   
21              22  Albania     80000  ALB0005  Non-OECD    Albania         2   
22              23  Albania     80000  ALB0005  Non-OECD    Albania         2   
23              24  Albania     80000  ALB0005  Non-OECD    Albania         2   
24              25  Albania     80000  ALB0005  Non-OECD    Albania         2   
25              26  Albania     80000  ALB0005  Non-OECD    Albania         2   
26              27  Albania     80000  ALB0005  Non-OECD    Albania         2   
27              28  Albania     80000  ALB0005  Non-OECD    Albania         2   
28              29  Albania     80000  ALB0005  Non-OECD    Albania         2   
29              30  Albania     80000  ALB0005  Non-OECD    Albania         2   
...            ...      ...       ...      ...       ...        ...       ...   
485460      485461  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485461      485462  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485462      485463  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485463      485464  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485464      485465  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485465      485466  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485466      485467  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485467      485468  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485468      485469  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485469      485470  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485470      485471  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485471      485472  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485472      485473  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485473      485474  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485474      485475  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485475      485476  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485476      485477  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485477      485478  Vietnam   7040000  

In [4]:
pisa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 485490 entries, 0 to 485489
Columns: 636 entries, Unnamed: 0 to VER_STU
dtypes: float64(250), int64(18), object(368)
memory usage: 2.3+ GB


In [5]:
pisa_dict

Unnamed: 0                                                  x
0           CNT                           Country code 3-character
1      SUBNATIO  Adjudicated sub-region code 7-digit code (3-di...
2       STRATUM  Stratum ID 7-character (cnt + region ID + orig...
3          OECD                                       OECD country
4            NC                       National Centre 6-digit Code
5      SCHOOLID  School ID 7-digit (region ID + stratum ID + 3-...
6       STIDSTD                                         Student ID
7       ST01Q01                                International Grade
8       ST02Q01                           National Study Programme
9       ST03Q01                                      Birth - Month
10      ST03Q02                                        Birth -Year
11      ST04Q01                                             Gender
12      ST05Q01                                   Attend <ISCED 0>
13      ST06Q01                                   Age at <ISCED 1>
14      ST07Q01                                 Repeat - <ISCED 1>
15      ST07Q02                                 Repeat - <ISCED 2>
16      ST07Q03                                 Repeat - <ISCED 3>
17      ST08Q01                          Truancy - Late for School
18      ST09Q01                    Truancy - Skip whole school day
19     ST115Q01           Truancy - Skip classes within school day
20      ST11Q01                                   At Home - Mother
21      ST11Q02                                   At Home - Father
22      ST11Q03                                 At Home - Brothers
23      ST11Q04                                  At Home - Sisters
24      ST11Q05                             At Home - Grandparents
25      ST11Q06                                   At Home - Others
26      ST13Q01                          Mother<Highest Schooling>
27      ST14Q01            Mother Qualifications - <ISCED level 6>
28      ST14Q02           Mother Qualifications - <ISCED level 5A>
29      ST14Q03           Mother Qualifications - <ISCED level 5B>
..          ...                                                ...
605    W_FSTR55           FINAL STUDENT REPLICATE BRR-FAY WEIGHT55
606    W_FSTR56           FINAL STUDENT REPLICATE BRR-FAY WEIGHT56
607    W_FSTR57           FINAL STUDENT REPLICATE BRR-FAY WEIGHT57
608    W_FSTR58           FINAL STUDENT REPLICATE BRR-FAY WEIGHT58
609    W_FSTR59           FINAL STUDENT REPLICATE BRR-FAY WEIGHT59
610    W_FSTR60           FINAL STUDENT REPLICATE BRR-FAY WEIGHT60
611    W_FSTR61           FINAL STUDENT REPLICATE BRR-FAY WEIGHT61
612    W_FSTR62           FINAL STUDENT REPLICATE BRR-FAY WEIGHT62
613    W_FSTR63           FINAL STUDENT REPLICATE BRR-FAY WEIGHT63
614    W_FSTR64           FINAL STUDENT REPLICATE BRR-FAY WEIGHT64
615    W_FSTR65           FINAL STUDENT REPLICATE BRR-FAY WEIGHT65
616    W_FSTR66           FINAL STUDENT REPLICATE BRR-FAY WEIGHT66
617    W_FSTR67           FINAL STUDENT REPLICATE BRR-FAY WEIGHT67
618    W_FSTR68           FINAL STUDENT REPLICATE BRR-FAY WEIGHT68
619    W_FSTR69           FINAL STUDENT REPLICATE BRR-FAY WEIGHT69
620    W_FSTR70           FINAL STUDENT REPLICATE BRR-FAY WEIGHT70
621    W_FSTR71           FINAL STUDENT REPLICATE BRR-FAY WEIGHT71
622    W_FSTR72           FINAL STUDENT REPLICATE BRR-FAY WEIGHT72
623    W_FSTR73           FINAL STUDENT REPLICATE BRR-FAY WEIGHT73
624    W_FSTR74           FINAL STUDENT REPLICATE BRR-FAY WEIGHT74
625    W_FSTR75           FINAL STUDENT REPLICATE BRR-FAY WEIGHT75
626    W_FSTR76           FINAL STUDENT REPLICATE BRR-FAY WEIGHT76
627    W_FSTR77           FINAL STUDENT REPLICATE BRR-FAY WEIGHT77
628    W_FSTR78           FINAL STUDENT REPLICATE BRR-FAY WEIGHT78
629    W_FSTR79           FINAL STUDENT REPLICATE BRR-FAY WEIGHT79
630    W_FSTR80           FINAL STUDENT REPLICATE BRR-FAY WEIGHT80
631    WVARSTRR           RANDOMIZED FINAL VARIANCE STRATUM (1-80)
632    VAR_UNIT                    RANDOMLY ASSIGNED

In [6]:
# Check for null values in relevant columns
columns = ['CNT', 'STIDSTD', 'ST01Q01', 'ST04Q01', 'ST13Q01', 
            'ST17Q01', 'ST28Q01', 'ST57Q01', 'ST57Q02', 'ST57Q03', 
            'ST57Q04', 'ST57Q05', 'MMINS', 'LMINS', 'SMINS',
            'PV1MATH', 'PV2MATH', 'PV3MATH', 'PV4MATH', 'PV5MATH',  
            'PV1READ', 'PV2READ',  'PV3READ', 'PV4READ', 'PV5READ',
            'PV1SCIE', 'PV2SCIE', 'PV3SCIE', 'PV4SCIE', 'PV5SCIE']

for x in columns:
    print(x, pisa[x].isna().sum())

CNT 0
STIDSTD 0
ST01Q01 0
ST04Q01 0
ST13Q01 27511
ST17Q01 42229
ST28Q01 11725
ST57Q01 184123
ST57Q02 215682
ST57Q03 201677
ST57Q04 205833
ST57Q05 195988
MMINS 202187
LMINS 202624
SMINS 214576
PV1MATH 0
PV2MATH 0
PV3MATH 0
PV4MATH 0
PV5MATH 0
PV1READ 0
PV2READ 0
PV3READ 0
PV4READ 0
PV5READ 0
PV1SCIE 0
PV2SCIE 0
PV3SCIE 0
PV4SCIE 0
PV5SCIE 0


In [7]:
pisa['CNT'].value_counts()

Mexico                      33806
Italy                       31073
Spain                       25313
Canada                      21544
Brazil                      19204
Australia                   14481
United Kingdom              12659
United Arab Emirates        11500
Switzerland                 11229
Qatar                       10966
Colombia                     9073
Finland                      8829
Belgium                      8597
Denmark                      7481
Jordan                       7038
Chile                        6856
Thailand                     6606
Japan                        6351
Chinese Taipei               6046
Peru                         6035
Slovenia                     5911
Argentina                    5908
Kazakhstan                   5808
Portugal                     5722
Indonesia                    5622
Singapore                    5546
Macao-China                  5335
Czech Republic               5327
Uruguay                      5315
Bulgaria      

In [8]:
pisa.duplicated().sum()

0

<a id='issues1'></a>
### Identified Issues
Here we can see some of the issues noticed in the dataframes. These quality issues will be cleaned in the following section.

1. Narrow down the dataframe by only looking at variables of interest since there are more variables than needed
2. Make the variable names readable since they are codes that can only be read with the dictionary
3. Create an average of all Math, Reading, and Science scores since all 5 scores for each are not necessary
4. Missing values in relevant columns ('Mother Highest Schooling', 'Father Highest Schooling', 'How many books at home', 'Out-of-School Study Time - Homework', 'Out-of-School Study Time - Guided Homework', 'Out-of-School Study Time - Personal Tutor','Out-of-School Study Time - Commercial Company', 'Out-of-School Study Time - With Parent', 'Learning time (minutes per week)- Mathematics', 'Learning time (minutes per week)  - test language', 'Learning time (minutes per week) - Science')
5. Create a total for Out-of-School Study Time and Learning time (minutes per week)
6. Single countries have various entries in the country column (Florida (USA), Massachusetts (USA), Connecticut (USA) and United States of America)


<a id='cleaning1'></a>
### First Cleaning

In [9]:
# Create a copy to preserve the original dataframe
pisa_clean = pisa.copy()

### Narrow down the dataframe by only looking at variables of interest

There are 636 columns available in the original dataframe, which is far more than what is necessary. We will rather look at just the following variables.

In [10]:
pisa_clean = pisa_clean[['CNT', 'ST01Q01', 'ST04Q01', 'ST13Q01', 
                         'ST17Q01', 'ST28Q01', 'ST57Q01', 'ST57Q02', 'ST57Q03', 
                         'ST57Q04', 'ST57Q05', 'MMINS', 'LMINS', 'SMINS',
                         'PV1MATH', 'PV2MATH', 'PV3MATH', 'PV4MATH', 'PV5MATH',  
                         'PV1READ', 'PV2READ',  'PV3READ', 'PV4READ', 'PV5READ',
                         'PV1SCIE', 'PV2SCIE', 'PV3SCIE', 'PV4SCIE', 'PV5SCIE']]       

In [11]:
# Check to see if the columns are correctly reduced
list(pisa_clean)

['CNT',
 'ST01Q01',
 'ST04Q01',
 'ST13Q01',
 'ST17Q01',
 'ST28Q01',
 'ST57Q01',
 'ST57Q02',
 'ST57Q03',
 'ST57Q04',
 'ST57Q05',
 'MMINS',
 'LMINS',
 'SMINS',
 'PV1MATH',
 'PV2MATH',
 'PV3MATH',
 'PV4MATH',
 'PV5MATH',
 'PV1READ',
 'PV2READ',
 'PV3READ',
 'PV4READ',
 'PV5READ',
 'PV1SCIE',
 'PV2SCIE',
 'PV3SCIE',
 'PV4SCIE',
 'PV5SCIE']

### Make the variable names readable

In the original dataframe, the variable names are all codes and unreadable without the dictionary. Therefore, we will change them to accurately reflect the information within the column.

In [12]:
# Rename columns
pisa_clean.rename({'CNT': 'Country', 
                   'ST01Q01': 'International Grade', 
                   'ST04Q01': 'Gender', 
                   'ST13Q01': 'Mother<Highest Schooling>', 
                   'ST17Q01': 'Father<Highest Schooling>', 
                   'ST28Q01': 'How many books at home', 
                   'ST57Q01': 'Out-of-School Study Time - Homework',
                   'ST57Q02': 'Out-of-School Study Time - Guided Homework',
                   'ST57Q03': 'Out-of-School Study Time - Personal Tutor',
                   'ST57Q04': 'Out-of-School Study Time - Commercial Company',
                   'ST57Q05': 'Out-of-School Study Time - With Parent',
                   'MMINS': 'Learning time (minutes per week)- <Mathematics>',
                   'LMINS': 'Learning time (minutes per week)  - <test language>',
                   'SMINS': 'Learning time (minutes per week) - <Science>'}, axis='columns', inplace=True)

In [13]:
# Check to see if the columns are correctly renamed
list(pisa_clean)

['Country',
 'International Grade',
 'Gender',
 'Mother<Highest Schooling>',
 'Father<Highest Schooling>',
 'How many books at home',
 'Out-of-School Study Time - Homework',
 'Out-of-School Study Time - Guided Homework',
 'Out-of-School Study Time - Personal Tutor',
 'Out-of-School Study Time - Commercial Company',
 'Out-of-School Study Time - With Parent',
 'Learning time (minutes per week)- <Mathematics>',
 'Learning time (minutes per week)  - <test language>',
 'Learning time (minutes per week) - <Science>',
 'PV1MATH',
 'PV2MATH',
 'PV3MATH',
 'PV4MATH',
 'PV5MATH',
 'PV1READ',
 'PV2READ',
 'PV3READ',
 'PV4READ',
 'PV5READ',
 'PV1SCIE',
 'PV2SCIE',
 'PV3SCIE',
 'PV4SCIE',
 'PV5SCIE']

### Create an average of all Math, Reading, and Science scores

For each of the Math, Reading, and Science scores, there are 5 scores provided. The variation between these scores is not needed, so rather we will look at the average for each of these major categories and remove the individual scores.

In [14]:
# Create average of Math scores
pisa_clean['Average Math Score']    = (pisa_clean['PV1MATH'] + pisa_clean['PV2MATH'] + pisa_clean['PV3MATH'] + pisa_clean['PV4MATH'] + pisa_clean['PV5MATH']) / 5

# Create average of Reading scores
pisa_clean['Average Reading Score'] = (pisa_clean['PV1READ'] + pisa_clean['PV2READ'] + pisa_clean['PV3READ'] + pisa_clean['PV4READ'] + pisa_clean['PV5READ']) / 5

# Create average of Science scores
pisa_clean['Average Science Score'] = (pisa_clean['PV1SCIE'] + pisa_clean['PV2SCIE'] + pisa_clean['PV3SCIE'] + pisa_clean['PV4SCIE'] + pisa_clean['PV5SCIE']) / 5

# Create average score
pisa_clean['Average Total Score']   = (pisa_clean['Average Math Score'] + pisa_clean['Average Reading Score'] + pisa_clean['Average Science Score'] +pisa_clean['Average Math Score']) / 3

In [15]:
# Remove the individual scores for Math, Reading, and Science
pisa_clean.drop(columns=['PV1MATH',  'PV2MATH', 'PV3MATH', 'PV4MATH', 'PV5MATH',   
                         'PV1READ', 'PV2READ',  'PV3READ', 'PV4READ',  'PV5READ',
                         'PV1SCIE',  'PV2SCIE', 'PV3SCIE',  'PV4SCIE', 'PV5SCIE'], inplace=True)

In [16]:
# Test to see accuracy of the Math score
print(pisa_clean['Average Math Score'].describe())
print('-'*22)

# Test to see accuracy of the Reading score
print(pisa_clean['Average Reading Score'].describe())
print('-'*22)

# Test to see accuracy of the Science score
print(pisa_clean['Average Science Score'].describe())

# Test to see accuracy of the Total score
print(pisa_clean['Average Total Score'].describe())

count    485490.000000
mean        469.651234
std         100.786610
min          54.767080
25%         396.019620
50%         465.734520
75%         540.123060
max         903.107960
Name: Average Math Score, dtype: float64
----------------------
count    485490.000000
mean        472.006964
std          98.863310
min           6.445400
25%         405.044200
50%         475.477980
75%         542.831195
max         849.359740
Name: Average Reading Score, dtype: float64
----------------------
count    485490.000000
mean        475.808094
std          97.998470
min          25.158540
25%         405.762800
50%         475.512860
75%         546.381920
max         857.832900
Name: Average Science Score, dtype: float64
count    485490.000000
mean        629.039175
std         128.839689
min         105.496473
25%         536.312167
50%         627.318087
75%         721.099990
max        1123.882273
Name: Average Total Score, dtype: float64


In [17]:
pisa_clean.sample(10)

Country  International Grade  Gender  \
428677                Serbia                    9  Female   
35654              Australia                   10  Female   
335572                Mexico                   10  Female   
83184                 Canada                   10  Female   
439277              Slovenia                   10  Female   
281751                 Japan                   10  Female   
16000   United Arab Emirates                   10    Male   
455192              Thailand                   10    Male   
21047              Argentina                   10    Male   
103152           Switzerland                   10  Female   

       Mother<Highest Schooling> Father<Highest Schooling>  \
428677     <ISCED level 3B, 3C>      <ISCED level 3B, 3C>    
35654          <ISCED level 3A>          <ISCED level 3A>    
335572     <ISCED level 3B, 3C>           <ISCED level 2>    
83184                        NaN         <ISCED level 3A>    
439277     <ISCED level 3B, 3C>      <ISCED level 3B, 3C>    
281751     <ISCED level 3B, 3C>          <ISCED level 3A>    
16000          <ISCED level 3A>          <ISCED level 3A>    
455192          <ISCED level 1>           <ISCED level 1>    
21047           <ISCED level 2>          <ISCED level 3A>    
103152     <ISCED level 3B, 3C>          <ISCED level 3A>    

       How many books at home  Out-of-School Study Time - Homework  \
428677          26-100 books                                   7.0   
35654            11-25 books                                   4.0   
335572                    NaN                                  4.0   
83184             0-10 books                                   0.0   
439277          26-100 books                                   6.0   
281751          26-100 books                                   1.0   
16000          201-500 books                                   3.0   
455192           11-25 books                                   NaN   
21047            11-25 books                                   NaN   
103152         101-200 books                                  14.0   

        Out-of-School Study Time - Guided Homework  \
428677                                         0.0   
35654                                          3.0   
335572                                         2.0   
83184                                          0.0   
439277                                         4.0   
281751                                         0.0   
16000                                          0.0   
455192                                         NaN   
21047                                          NaN   
103152                                         7.0   

        Out-of-School Study Time - Personal Tutor  \
428677                                        0.0   
35654                                         1.0   
335572                                        NaN   
83184                                         0.0   
439277                                        0.0   
281751                                        0.0   
16000                                         0.0   
455192                                        NaN   
21047                                         NaN   
103152                                        0.0   

        Out-of-School Study Time - Commercial Company  \
428677                                            0.0   
35654                                             0.0   
335572                                            NaN   
83184                                             0.0   
439277                                            0.0   
281751                                            0.0   
16000                                             0.0   
455192                                            NaN   
21047                                             NaN   
103152                                            0.0   

        Out-of-School Study Time - With Parent  \
428677                                

### 4. Missing values in relevant columns ('Mother Highest Schooling', 'Father Highest Schooling', 'How many books at home', 'Out-of-School Study Time - Homework', 'Out-of-School Study Time - Guided Homework', 'Out-of-School Study Time - Personal Tutor','Out-of-School Study Time - Commercial Company', 'Out-of-School Study Time - With Parent', 'Learning time (minutes per week)- Mathematics', 'Learning time (minutes per week)  - test language', 'Learning time (minutes per week) - Science')

Due to the amount of rows in this dataframe, we can comfortably remove the rows with missing values.

In [18]:

pisa_clean.replace(["NaN"], np.nan, inplace = True)
pisa_clean.dropna(inplace = True)

In [19]:
# Check to see if null values correctly dropped
pisa_clean

Country  International Grade  Gender  \
3       Albania                    9  Female   
6       Albania                   10  Female   
13      Albania                   10  Female   
17      Albania                   10    Male   
18      Albania                   10  Female   
27      Albania                   10    Male   
28      Albania                   10  Female   
46      Albania                    9  Female   
53      Albania                   10    Male   
55      Albania                    9    Male   
57      Albania                    9    Male   
58      Albania                   10    Male   
61      Albania                   10  Female   
62      Albania                   10  Female   
63      Albania                    9  Female   
66      Albania                    9    Male   
73      Albania                    9  Female   
75      Albania                    9    Male   
84      Albania                   10  Female   
86      Albania                   10  Female   
87      Albania                   10    Male   
89      Albania                   10  Female   
92      Albania                   10    Male   
94      Albania                   10    Male   
96      Albania                    9    Male   
97      Albania                   10    Male   
103     Albania                   10  Female   
108     Albania                   10  Female   
123     Albania                   10    Male   
124     Albania                   10    Male   
...         ...                  ...     ...   
485378  Vietnam                   10  Female   
485383  Vietnam                   10    Male   
485393  Vietnam                   10    Male   
485398  Vietnam                   10  Female   
485399  Vietnam                   10  Female   
485400  Vietnam                   10  Female   
485407  Vietnam                   10    Male   
485411  Vietnam                   10    Male   
485412  Vietnam                   10    Male   
485417  Vietnam                   10    Male   
485419  Vietnam                   10  Female   
485424  Vietnam                   10  Female   
485425  Vietnam                   10    Male   
485427  Vietnam                   10  Female   
485429  Vietnam                   10    Male   
485431  Vietnam                   10  Female   
485434  Vietnam                   10  Female   
485435  Vietnam                   10    Male   
485437  Vietnam                   10  Female   
485440  Vietnam                   10    Male   
485442  Vietnam                   10    Male   
485444  Vietnam                   10  Female   
485446  Vietnam                   10  Female   
485453  Vietnam                   10  Female   
485458  Vietnam                   10    Male   
485460  Vietnam                   10    Male   
485463  Vietnam                   10    Male   
485469  Vietnam                   10  Female   
485472  Vietnam                   10    Male   
485475  Vietnam                   10    Male   

                    Mother<Highest Schooling>  \
3                       <ISCED level 3B, 3C>    
6                       <ISCED level 3B, 3C>    
13                          <ISCED level 3A>    
17                           <ISCED level 2>    
18                          <ISCED level 3A>    
27                          <ISCED level 3A>    
28                           <ISCED level 2>    
46                           <ISCED level 2>    
53                          <ISCED level 3A>    
55                           <ISCED level 2>    
57      She did not complete <ISCED level 1>    
58                           <ISCED level 2>    
61                          <ISCED level 3A>    
62                           <ISCED level 2>    
63                           <ISCED level 2>    
66                          <ISCED level 3A>    
73                          <ISCED level 3A>    
75                          <ISCED level 3A>    
84                           <ISCED level 2>    
86                          <ISCED level 3A>    
87

In [20]:
# Confirm with smaller shape of dataframe
pisa_clean.shape

(189773, 18)

### 5. Create a total for Out-of-School Study Time and Learning time (minutes per week)



In [21]:
# Create total of Out-of-School Study Time
pisa_clean['Out-of-School Study Time - Total'] = pisa_clean['Out-of-School Study Time - Homework'] + pisa_clean['Out-of-School Study Time - Guided Homework'] + pisa_clean['Out-of-School Study Time - Personal Tutor'] + pisa_clean['Out-of-School Study Time - Commercial Company'] + pisa_clean['Out-of-School Study Time - With Parent']

# Create total of Learning time (minutes per week)
pisa_clean['Learning time (minutes per week) - Total'] = pisa_clean['Learning time (minutes per week)- <Mathematics>'] + pisa_clean['Learning time (minutes per week)  - <test language>'] + pisa_clean['Learning time (minutes per week) - <Science>'] 
                                                                                                                 

In [22]:
list(pisa_clean)

['Country',
 'International Grade',
 'Gender',
 'Mother<Highest Schooling>',
 'Father<Highest Schooling>',
 'How many books at home',
 'Out-of-School Study Time - Homework',
 'Out-of-School Study Time - Guided Homework',
 'Out-of-School Study Time - Personal Tutor',
 'Out-of-School Study Time - Commercial Company',
 'Out-of-School Study Time - With Parent',
 'Learning time (minutes per week)- <Mathematics>',
 'Learning time (minutes per week)  - <test language>',
 'Learning time (minutes per week) - <Science>',
 'Average Math Score',
 'Average Reading Score',
 'Average Science Score',
 'Average Total Score',
 'Out-of-School Study Time - Total',
 'Learning time (minutes per week) - Total']

### 6. Single countries have various entries in the country column (Florida (USA), Massachusetts (USA), Connecticut (USA) and United States of America)

In [23]:
pisa_clean.replace(['Connecticut (USA)', 'Florida (USA)', 'Massachusetts (USA)'], 'United States of America', inplace=True)
pisa_clean.replace(['Perm(Russian)', 'Russian Federation', 'Perm(Russian Federation)'], 'Russia', inplace=True)

In [24]:
pisa_clean['Country'].value_counts()

Italy                       14482
Mexico                      11510
Spain                       11149
Canada                      10773
United Kingdom               6046
Switzerland                  5383
United States of America     5279
Australia                    5150
Brazil                       4553
Finland                      4502
United Arab Emirates         3616
Belgium                      3491
Denmark                      3474
Japan                        3413
Thailand                     3161
Singapore                    3005
Chinese Taipei               2960
Croatia                      2827
Poland                       2774
China-Shanghai               2723
Czech Republic               2722
Korea                        2566
Russia                       2543
Estonia                      2510
Ireland                      2448
Slovenia                     2369
Macao-China                  2333
Sweden                       2310
Luxembourg                   2286
Netherlands   

<a id='assessing2'></a>
### Final Assessment

In this section, we will do the preliminary visual and programmatic assessment of the pisa datasets to determine whether or not it holds any major quality or tidiness issues.

In [25]:
pisa_clean

Country  International Grade  Gender  \
3       Albania                    9  Female   
6       Albania                   10  Female   
13      Albania                   10  Female   
17      Albania                   10    Male   
18      Albania                   10  Female   
27      Albania                   10    Male   
28      Albania                   10  Female   
46      Albania                    9  Female   
53      Albania                   10    Male   
55      Albania                    9    Male   
57      Albania                    9    Male   
58      Albania                   10    Male   
61      Albania                   10  Female   
62      Albania                   10  Female   
63      Albania                    9  Female   
66      Albania                    9    Male   
73      Albania                    9  Female   
75      Albania                    9    Male   
84      Albania                   10  Female   
86      Albania                   10  Female   
87      Albania                   10    Male   
89      Albania                   10  Female   
92      Albania                   10    Male   
94      Albania                   10    Male   
96      Albania                    9    Male   
97      Albania                   10    Male   
103     Albania                   10  Female   
108     Albania                   10  Female   
123     Albania                   10    Male   
124     Albania                   10    Male   
...         ...                  ...     ...   
485378  Vietnam                   10  Female   
485383  Vietnam                   10    Male   
485393  Vietnam                   10    Male   
485398  Vietnam                   10  Female   
485399  Vietnam                   10  Female   
485400  Vietnam                   10  Female   
485407  Vietnam                   10    Male   
485411  Vietnam                   10    Male   
485412  Vietnam                   10    Male   
485417  Vietnam                   10    Male   
485419  Vietnam                   10  Female   
485424  Vietnam                   10  Female   
485425  Vietnam                   10    Male   
485427  Vietnam                   10  Female   
485429  Vietnam                   10    Male   
485431  Vietnam                   10  Female   
485434  Vietnam                   10  Female   
485435  Vietnam                   10    Male   
485437  Vietnam                   10  Female   
485440  Vietnam                   10    Male   
485442  Vietnam                   10    Male   
485444  Vietnam                   10  Female   
485446  Vietnam                   10  Female   
485453  Vietnam                   10  Female   
485458  Vietnam                   10    Male   
485460  Vietnam                   10    Male   
485463  Vietnam                   10    Male   
485469  Vietnam                   10  Female   
485472  Vietnam                   10    Male   
485475  Vietnam                   10    Male   

                    Mother<Highest Schooling>  \
3                       <ISCED level 3B, 3C>    
6                       <ISCED level 3B, 3C>    
13                          <ISCED level 3A>    
17                           <ISCED level 2>    
18                          <ISCED level 3A>    
27                          <ISCED level 3A>    
28                           <ISCED level 2>    
46                           <ISCED level 2>    
53                          <ISCED level 3A>    
55                           <ISCED level 2>    
57      She did not complete <ISCED level 1>    
58                           <ISCED level 2>    
61                          <ISCED level 3A>    
62                           <ISCED level 2>    
63                           <ISCED level 2>    
66                          <ISCED level 3A>    
73                          <ISCED level 3A>    
75                          <ISCED level 3A>    
84                           <ISCED level 2>    
86                          <ISCED level 3A>    
87

In [26]:
pisa_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 189773 entries, 3 to 485475
Data columns (total 20 columns):
Country                                                189773 non-null object
International Grade                                    189773 non-null int64
Gender                                                 189773 non-null object
Mother<Highest Schooling>                              189773 non-null object
Father<Highest Schooling>                              189773 non-null object
How many books at home                                 189773 non-null object
Out-of-School Study Time - Homework                    189773 non-null float64
Out-of-School Study Time - Guided Homework             189773 non-null float64
Out-of-School Study Time - Personal Tutor              189773 non-null float64
Out-of-School Study Time - Commercial Company          189773 non-null float64
Out-of-School Study Time - With Parent                 189773 non-null float64
Learning time (minutes per 

<a id='sav'></a>
## Store

In [27]:
# Store the clean DataFrame in a CSV file
pisa_clean.to_csv("pisa_df.csv", index=False, encoding='utf8')

<a id='conclusion'></a>
## Conclusion

<a id='references'></a>
## References